File: src/lib/llm.ts
/**
 * Wrapper minimal untuk Gemini API (REST, tanpa SDK tambahan).
 *
 * Aktif jika environment variable `GEMINI_API_KEY` terisi. Kalau tidak,
 * chatbot tetap jalan via rule-based dan grading esai akan return null.
 *
 * Dokumen: https://ai.google.dev/api/generate-content
 */

const API_BASE = "https://generativelanguage.googleapis.com/v1beta";

export function isLlmEnabled(): boolean {
  return !!process.env.GEMINI_API_KEY;
}

export function llmModel(): string {
  return process.env.GEMINI_MODEL || "gemini-2.5-flash";
}

type GenerateArgs = {
  systemInstruction: string;
  userMessage: string;
  temperature?: number;
  maxOutputTokens?: number;
  /** Jika diset, paksa output JSON sesuai schema. */
  responseSchema?: Record<string, unknown>;
  /** Untuk Gemini 2.5+: nonaktifkan thinking agar respons cepat & tak menghabiskan token. */
  disableThinking?: boolean;
};

export type GenerateResult = {
  text: string;
  blocked: boolean;
  json?: unknown;
};

/** Panggil generateContent endpoint Gemini. Throw jika error jaringan/auth. */
export async function generateContent(args: GenerateArgs): Promise<GenerateResult> {
  const key = process.env.GEMINI_API_KEY;
  if (!key) throw new Error("GEMINI_API_KEY belum diset");

  const model = llmModel();
  const generationConfig: Record<string, unknown> = {
    temperature: args.temperature ?? 0.4,
    maxOutputTokens: args.maxOutputTokens ?? 2048,
  };
  if (args.responseSchema) {
    generationConfig.responseMimeType = "application/json";
    generationConfig.responseSchema = args.responseSchema;
  }
  // Gemini 2.5+ default thinking. Untuk grading yang butuh JSON cepat,
  // matikan agar respons tidak terpotong oleh thinking tokens.
  if (args.disableThinking) {
    generationConfig.thinkingConfig = { thinkingBudget: 0 };
  }
  const body: Record<string, unknown> = {
    systemInstruction: { parts: [{ text: args.systemInstruction }] },
    contents: [{ role: "user", parts: [{ text: args.userMessage }] }],
    generationConfig,
  };

  const url = `${API_BASE}/models/${model}:generateContent?key=${key}`;
  const res = await fetch(url, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
    cache: "no-store",
  });

  if (!res.ok) {
    const errBody = await res.text().catch(() => "");
    throw new Error(`Gemini API ${res.status}: ${errBody.slice(0, 300)}`);
  }

  const data = (await res.json()) as {
    candidates?: Array<{
      content?: { parts?: Array<{ text?: string }> };
      finishReason?: string;
    }>;
    promptFeedback?: { blockReason?: string };
  };

  if (data.promptFeedback?.blockReason) {
    return { text: "[Permintaan diblokir oleh filter keamanan]", blocked: true };
  }

  const cand = data.candidates?.[0];
  if (cand?.finishReason === "SAFETY") {
    return {
      text: "Maaf, saya tidak dapat menjawab pertanyaan ini karena memuat konten yang tidak sesuai filter keamanan.",
      blocked: true,
    };
  }
  if (cand?.finishReason === "MAX_TOKENS") {
    // Bisa jadi thinking habis budget. Tetap kembalikan partial text + tandai blocked
    // agar caller bisa fallback.
    const partial = cand?.content?.parts?.map((p) => p.text ?? "").join("") ?? "";
    if (partial.length < 10) {
      return { text: "[Respons terpotong oleh batas token]", blocked: true };
    }
  }

  const text = cand?.content?.parts?.map((p) => p.text ?? "").join("") ?? "";
  let json: unknown = undefined;
  if (args.responseSchema) {
    try {
      json = JSON.parse(text);
    } catch {
      /* biarkan undefined */
    }
  }
  return { text, blocked: false, json };
}

File: lib/llm.py
"""Wrapper minimal untuk Gemini API (REST). Kompatibel dengan free tier.
Aktif jika GEMINI_API_KEY terisi di st.secrets atau env.
"""
from __future__ import annotations

import os
from typing import Any

import requests
import streamlit as st

API_BASE = "https://generativelanguage.googleapis.com/v1beta"


def _secret(key: str, default: str = "") -> str:
    if hasattr(st, "secrets"):
        try:
            v = st.secrets.get(key, None)
            if v:
                return str(v)
        except (FileNotFoundError, AttributeError, Exception):
            pass
    return os.environ.get(key, default)


def is_llm_enabled() -> bool:
    return bool(_secret("GEMINI_API_KEY"))


def llm_model() -> str:
    return _secret("GEMINI_MODEL") or "gemini-2.5-flash"


def _timeout_sec() -> float:
    try:
        return float(_secret("GEMINI_TIMEOUT_SEC") or 30)
    except (TypeError, ValueError):
        return 30.0


def generate_content(
    *,
    system_instruction: str,
    user_message: str,
    temperature: float = 0.4,
    max_output_tokens: int = 2048,
    response_schema: dict | None = None,
    disable_thinking: bool = False,
) -> dict[str, Any]:
    """
    Returns: {"text": str, "blocked": bool, "json": dict|None}
    Raises: RuntimeError on network/HTTP/timeout error.
    """
    key = _secret("GEMINI_API_KEY")
    if not key:
        raise RuntimeError("GEMINI_API_KEY belum diset")

    gen_config: dict[str, Any] = {
        "temperature": temperature,
        "maxOutputTokens": max_output_tokens,
    }
    if response_schema:
        gen_config["responseMimeType"] = "application/json"
        gen_config["responseSchema"] = response_schema
    if disable_thinking:
        gen_config["thinkingConfig"] = {"thinkingBudget": 0}

    body = {
        "systemInstruction": {"parts": [{"text": system_instruction}]},
        "contents": [{"role": "user", "parts": [{"text": user_message}]}],
        "generationConfig": gen_config,
    }
    url = f"{API_BASE}/models/{llm_model()}:generateContent?key={key}"

    try:
        res = requests.post(
            url,
            json=body,
            headers={"Content-Type": "application/json"},
            timeout=_timeout_sec(),
        )
    except requests.exceptions.Timeout as e:
        raise RuntimeError(f"Gemini API timeout ({_timeout_sec()}s)") from e
    except requests.exceptions.RequestException as e:
        raise RuntimeError(f"Gemini API jaringan error: {e}") from e

    if not res.ok:
        raise RuntimeError(f"Gemini API {res.status_code}: {res.text[:300]}")

    data = res.json()
    if data.get("promptFeedback", {}).get("blockReason"):
        return {"text": "[Diblokir oleh filter keamanan]", "blocked": True, "json": None}

    cand = (data.get("candidates") or [{}])[0]
    if cand.get("finishReason") == "SAFETY":
        return {"text": "Maaf, tidak dapat menjawab pertanyaan ini.", "blocked": True, "json": None}

    parts = (cand.get("content") or {}).get("parts") or []
    text = "".join(p.get("text", "") for p in parts)

    if cand.get("finishReason") == "MAX_TOKENS" and len(text) < 10:
        return {"text": "[Respons terpotong oleh batas token]", "blocked": True, "json": None}

    json_data = None
    if response_schema:
        try:
            import json as _json
            json_data = _json.loads(text)
        except (ValueError, TypeError):
            json_data = None

    return {"text": text, "blocked": False, "json": json_data}
